packages/go/stbernard/cmdrunner/cmdrunner.go

```diff
@@ -17,64 +17,122 @@
+// an executed command's output.
 //
-// If debug log level is set globally, command output will be combined and sent to os.Stderr.
-func Run(command string, args []string, path string, env environment.Environment, cmdModifiers ...func(*exec.Cmd)) error {
-	var (
-		exitErr error
+// The combined output buffer preserves the chronological order of both streams at time of collection from their
+// respective file descriptors.
+type ExecutionResult struct {
+	Command        string
+	Arguments      []string
+	Path           string
+	StandardOutput *bytes.Buffer
+	ErrorOutput    *bytes.Buffer
+	CombinedOutput *bytes.Buffer
+	ReturnCode     int
+}
+
+func (s *ExecutionResult) Error() string {
+	return "command execution failed: " + s.Command
+}
 
-		cmdstr       = command + " " + args[0]
-		cmd          = exec.Command(command, args...)
-		debugEnabled = level.GlobalAccepts(slog.LevelDebug)
+func newExecutionResult(command string, args []string) *ExecutionResult {
+	return &ExecutionResult{
+		Command:        command,
+		Arguments:      args,
+		StandardOutput: &bytes.Buffer{},
+		ErrorOutput:    &bytes.Buffer{},
+		CombinedOutput: &bytes.Buffer{},
+		ReturnCode:     0,
+	}
+}
+
+func prepareCommand(command string, args []string, path string, env environment.Environment) (*exec.Cmd, *ExecutionResult) {
+	var (
+		cmd    = exec.Command(command, args...)
+		result = newExecutionResult(command, args)
 	)
 
-	cmd.Env = env.Slice()
 	cmd.Dir = path
+	cmd.Env = env.Slice()
+	cmd.Stdout = io.MultiWriter(result.StandardOutput, result.CombinedOutput)
+	cmd.Stderr = io.MultiWriter(result.ErrorOutput, result.CombinedOutput)
 
-	// Default to mapping stdout directly to stdout
-	cmd.Stdout = os.Stdout
+	return cmd, result
+}
 
-	if debugEnabled {
-		cmd.Stderr = os.Stderr
-		cmdstr = command + " " + strings.Join(args, " ")
+func shortCommandString(command string, args []string) string {
+	if len(args) > 0 {
+		return command + " " + args[0]
 	}
 
-	// If we got any cmdModifiers, apply them in order
-	// This is often used for capturing Stdout and other modifications
-	if len(cmdModifiers) > 0 {
-		for _, modifier := range cmdModifiers {
-			modifier(cmd)
+	return command
+}
+
+// logCommand outputs command execution intent into the log with a short version of the command and its arguments. The
+// returned closure will emit the result of the executed command along with more detailed information including
+// elapsed run time to debug output.
+func logCommand(result *ExecutionResult) func() {
+	var (
+		commandStr = shortCommandString(result.Command, result.Arguments)
+		started    = time.Now()
+	)
+
+	slog.Info("exec", slog.String("command", commandStr))
+
+	return func() {
+		var (
+			formattedArgs = strings.Join(result.Arguments, " ")
+			elapsed       = time.Since(started)
+		)
+
+		if result.ReturnCode != 0 {
+			if _, err := io.Copy(os.Stderr, result.CombinedOutput); err != nil {
```

Comment:
I would encourage us to use `result.ErrorOutput()` instead here, since combined output will be more noisy in cases like `golangci-lint` outputting valid results (it will give a non-zero exit and put the json on stdout with additional information on stderr). Most standard commands are really good about proper stdout/stderr usage, especially anything that expects to run in CI environments